Add auto deploy to Firebase

chuckh opened the pull request:
This adds the auto deploy part. Does not address testing.

Getting this working maybe iterative process like auto deploying GitHub
Pages was.

Addresses auto deploy part of #583 

r: @robdodson @samccone

samccone commented on .travis.yml:
```diff
@@ -16,6 +16,7 @@ before_script:
   - sleep 3
   - npm i -g bower gulp
   - bower i
+  - npm install -g firebase-tools
```
move this up to line 17 :)

samccone replied:
also lets put a @version after all of them :)

chuckh replied:
@samccone like this?

```
before_script:
  - 'export DISPLAY=:99.0'
  - sh -e /etc/init.d/xvfb start
  - sleep 3
  - npm i -g firebase-tools @version
  - npm i -g bower gulp @version
  - bower i @version
```

samccone replied:
```
- npm i -g bower@x.x.x gulp@x.x.x firebase-tools@x.x.x
```

:)_

chuckh replied:
@samccone fixed. :)

samccone commented on travis-runner.sh:
```diff
@@ -5,13 +5,28 @@ if [ "$TRAVIS_BRANCH" = "master" ] && [ "$TRAVIS_PULL_REQUEST" = "false" ]
+  rm app/elements/routing.html.tmp && \
+  rm app/index.html.tmp && \
+  echo " DONE deploying to Firebase!"
 else
```
my rat called... he wants back his nest :cry: 

....


we need to figure out some way to make this better hahaha

chuckh replied:
I am open to suggestions. This is just what worked for me to do it manually translated to work with Travis CI.

samccone replied:
oh yeah totally, I need to think about how we can make this even just a touch more declarative...

splitting these deploy actions up into functions might be a decent first step
http://tldp.org/LDP/abs/html/functions.html

:)

chuckh replied:
@samccone like this? We just have to insure that they run in sequence not parallel.

```bash
#!/bin/bash
set -o pipefail

if [ "$TRAVIS_BRANCH" = "master" ] && [ "$TRAVIS_PULL_REQUEST" = "false" ]
then
  git config --global user.email "samccone@gmail.com" && \
  git config --global user.name "auto deployer" && \

  deploy_ghpages () {
    echo "Deploying to GitHub Pages!" && \
    sed -i.tmp "s/\/\/ app.baseUrl = '\/polymer-starter-kit/app.baseUrl = '\/polymer-starter-kit/" app/scripts/app.js && \
    bower i && \
    gulp build-deploy-gh-pages && \
    echo " Undoing Changes to PSK for GitHub Pages" && \
    cp app/scripts/app.js.tmp app/scripts/app.js && \
    rm app/scripts/app.js.tmp && \
    echo " DONE deploying to Github Pages!" && \
  }

  deploy_firebase () {
    echo "Deploying to Firebase! (https://polymer-starter-kit.firebaseapp.com)" && \
    echo " Making Changes to PSK for Firebase" && \
    sed -i.tmp 's/<!-- Chrome for Android theme color -->/<base href="\/">\'$'\n<!-- Chrome for Android theme color -->/g' app/index.html && \
    sed -i.tmp "s/hashbang: true/hashbang: false/" app/elements/routing.html && \
    echo " Starting Build Process for Firebase Changes" && \
    gulp && \
    echo " Starting Deploy Process to Firebaseapp.com Server -- polymer-starter-kit.firebaseapp.com" && \
    firebase deploy --token "${FIREBASE_TOKEN}" && \
    echo " Undoing Changes to PSK for Firebase" && \
    cp app/index.html.tmp app/index.html && \
    cp app/elements/routing.html.tmp app/elements/routing.html && \
    rm app/elements/routing.html.tmp && \
    rm app/index.html.tmp && \
    echo " DONE deploying to Firebase!"  && \
  }

  deploy_ghpages  && \
  deploy_firebase && \
  
else
  npm run lint
  npm test
fi
```

chuckh commented:
Removing version numbers (@x.x.x) worked, Travis CI now passes ok.

samccone commented:
hmmmmmm unexpected ha, do you have a link to the error?

samccone commented on firebase.json:
```diff
@@ -0,0 +1,13 @@
+{
```
I wonder if having this file in our project will cause issues for others when they go to deploy to firebase...

thoughts?

robdodson replied:
yeah good point. I wonder if we could have travis run the commands for the CLI and pass in the correct arguments.

mbleigh replied:
If you leave the "firebase" key out of the file you can pass -f when you
deploy and supply it.

On Sat, Dec 12, 2015, 9:58 AM Rob Dodson <notifications@github.com> wrote:

> In firebase.json
> <https://github.com/PolymerElements/polymer-starter-kit/pull/584#discussion_r47434699>
> :
>
> > @@ -0,0 +1,13 @@
> > +{
>
> yeah good point. I wonder if we could have travis run the commands for the
> CLI and pass in the correct arguments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PolymerElements/polymer-starter-kit/pull/584/files#r47434699>
> .
>

addyosmani commented:
I have similar concerns about the json file. Maybe @mbleigh may be able to confirm?

chuckh commented:
One option might be name it to firebase-psk.json, then rename it firebase.json before doing firebase deploy and then renaming it back.

chuckh commented on travis-runner.sh:
```diff
@@ -5,13 +5,28 @@ if [ "$TRAVIS_BRANCH" = "master" ] && [ "$TRAVIS_PULL_REQUEST" = "false" ]
+  echo "Deploying to GitHub Pages!" && \
   sed -i.tmp "s/\/\/ app.baseUrl = '\/polymer-starter-kit/app.baseUrl = '\/polymer-starter-kit/" app/scripts/app.js && \
-  rm app/scripts/app.js.tmp && \
   bower i && \
```
@samccone can we remove this `bower i &&\` since we are already doing it in .travis.yml line 18? https://github.com/PolymerElements/polymer-starter-kit/pull/584/files#diff-354f30a63fb0907d4ad57269548329e3L18

samccone replied:
yep we can :)

mbleigh commented:
FYI: I'm not sure if it's landed in production Traivs yet, but Firebase has official deploy support merged. See https://github.com/travis-ci/dpl/pull/358

chuckh commented:
I updated this pull request with the following:

1. Removed `"firebase": "polymer-starter-kit"` line from firebase.json
2. Changed travis-runner.sh to use firebase deploy -f `firebase deploy -f polymer-starter-kit --non-interactive --token "${FIREBASE_TOKEN}" && \`
3. Created functions for `deploy_ghpages` and `deploy_firebase` in travis-runner.sh
4. Removed extra `bower i` in travis-runner.sh

chuckh commented:
@robdodson one thought on copying the firebase.json from somewhere else, is to put it in the /docs folder as an example for the Firebase recipe and for copying for travis auto deploy to Firebase. This way it does not interfere with doing `firebase init` for new app based on PSK.

mbleigh commented:
That sounds like a reasonable solution to me, although having "public" and "rewrites" already set up is nice with the current setup.

chuckh commented:
I updated this pull request with the following:

1. Moved `app/firebase.json` to `app/docs/firebase.json`
2. Updated deploy firebase recipe to reflect this change
3. Added logic to copy `app/docs/firebase.json` to `app/firebase.json` during deploying to Firebase, then remove once done.

I think I have everyones recommended changes incorporated.   @robdodson moving firebase.json to docs folder I believe is a happy medium between having an extra file and having to copy file from another site.

robdodson commented:
Super bash newb question... are all the `&& \` bits necessary? If we removed the echos could we just run each line one after the other without all that?

samccone commented:
"&&" is used to chain commands together, such that the next command is run if and only if the preceding command exited without errors (or, more accurately, exits with a return code of 0).

newlines would not give us this behavior

mbleigh commented:
Could use `set -e` to make the script exit if any command fails maybe?

http://www.davidpashley.com/articles/writing-robust-shell-scripts/#id2382181

robdodson commented:
ah I see. Yeah I believe we're using that in our github pages script here (https://github.com/yeoman/generator-polymer/blob/master/gh/templates/gp.sh#L1)

Maybe we could try the same to tidy up this file a bit?

samccone commented:
for sure!

chuckh commented:
Changed to use `#!/bin/bash -e` and removed `&& \`s.  Changed echos to comments.

robdodson commented:
This is looking much better to me. Thanks for hanging in there @chuckh!
I think I'm going to say this LGTM. Maybe @samccone or @addyosmani can give the final sign off.

chuckh commented:
Ideally we land this pr and work out any kinks with travis before landing #614. @samccone / @addyosmani are you ready to land this?

chuckh commented:
I have rebased and squashed commits. All checks have passed. PTAL @samccone.  The pr related changes are in:

1. .travis.yml (line 19)
2. travis-runner.sh
3. docs/firebase.json
3. dos/deploy-to-firebase-pretty-urls.md

robdodson commented:
This looks like it may be rewriting a number of past commits. It may be better to pull the changes out and do a fresh PR

chuckh commented:
@robdodson will do. Creating new pr now. Will close this one once it is done.

chuckh commented:
PR #641 replaces this pr.
